controller.py

```diff
@@ -215,6 +256,55 @@ def search(self, location, live=True):
             )
             return Response(body, 200, headers)
 
+    def libraries(self):
+        # Return a specific set of information about all libraries; this generates the library list in the admin interface,
+        libraries = []
+        all = self._db.query(Library).order_by(Library.name)
+        for library in all:
+            uuid = library.internal_urn.split("uuid:")[1]
+            libraries += [dict(
+                    id=library.id,
+                    name=library.name,
+                    short_name=library.short_name,
+                    uuid=uuid,
+                    registry_stage=library.registry_stage,
+                    library_stage=library.library_stage,
+                )]
+        return dict(libraries=libraries)
+
+    def library_details(self, uuid):
+        # Return complete information about one specific library; this is triggered when an admin opens the detail page for a library.
+        library = self.library_for_request(uuid)
+        if isinstance(library, ProblemDetail):
+            return library
+        library_info = dict(
+            name=library.name,
+            short_name=library.short_name,
+            authentication_url=library.authentication_url,
+            online_registration=library.online_registration,
+            description=library.description,
+            timestamp=library.timestamp,
+            internal_urn=library.internal_urn,
+            library_stage=library._library_stage,
+            opds_url=library.opds_url,
+            registry_stage=library.registry_stage,
+            web_url=library.web_url,
+        )
+        return library_info
+
+    def edit_registration(self):
+        # Edit a specific library's registry_stage and library_stage based on information which an admin has submitted in the interface.
+        uuid = flask.request.form.get("uuid")
+        library = self.library_for_request(uuid)
+        if isinstance(library, ProblemDetail):
+            return library
+        registry_stage = flask.request.form.get("Registry Stage")
+        library_stage = flask.request.form.get("Library Stage")
+
+        library._library_stage = library_stage
```

Comment:
This is giving me a 500 error if I change library stage for a production library - "This library is already in production, only the registry can take it out of production."

Should the library stage even be editable by the registry admin?

Reply:
Yes, library stage should be editable by the registry admin.  I'm not getting the error--do you have specific steps to reproduce (beyond just setting the library stage to production and then trying to change it to something else)?

Reply:
You have to set both library and registry stage to production, then try to change the library stage but not the registry stage.

Why should library stage be editable?

Reply:
Library stage is the stage the library itself wants to be in, and should come from what the library set on the circ manager. Maybe the registry admin would want to fix it if the library was having trouble, but in general the library should change it on the circ manager and re-register. The registry stage always needs to be updated by registry admins. 

I tried this on Friday right before I wrote the comment.

Reply:
I think the point would be to let the admin fix it if the library was having trouble.  (The user base is comprised of you and Leonard; it's not like we have to worry about rogue admins going in and messing up the library stage for no good reason!)

Would you mind trying it again?  I definitely played around with the relevant code on Friday, but I don't remember exactly what time...

Reply:
The main case I was thinking of here is where the circulation manager just shuts down. (This has happened.) At that point we need to change both states to 'cancelled' and the normal rules about what's allowed don't apply.

But I agree that in most cases it doesn't make sense for the registry to modify the library_stage.

Reply:
Tried again and it is working now.